Kudu.Services/FetchHelpers/OneDriveHelper.cs

```diff
@@ -0,0 +1,519 @@
+
+                result.Cursor = next;
+                return result;
+            }
+        }
+
+        private ChangesResult GetChanges(OneDriveModel.OneDriveItemCollection items, Dictionary<string, OneDriveModel.ItemInfo> ids, string rootUri)
+        {
+            ChangesResult result = new ChangesResult();
+
+            foreach (var item in items.value)
+            {
+                ids[item.id] = new OneDriveModel.ItemInfo
+                {
+                    name = item.name,
+                    parentId = item.parentReference.id
+                };
+
+                var path = GetPath(ids, item);
+                if (item.deleted != null)
+                {
+                    result.DeletionChanges.Add(new OneDriveModel.OneDriveChange { Path = path, IsDeleted = true });
+                }
+                else
+                {
+                    var change = new OneDriveModel.OneDriveChange
+                    {
+                        ContentUri = string.Format(CultureInfo.InvariantCulture, "{0}/items/{1}", rootUri, item.id),
+                        Path = path,
+                        IsFile = item.file != null,
+                        LastModifiedUtc = item.lastModifiedDateTime.ToUniversalTime()
+                    };
+
+                    if (change.IsFile)
+                    {
+                        result.FileChanges.Add(change);
+                    }
+                    else
+                    {
+                        result.DirectoryChanges.Add(change);
+                    }
+                }
+            }
+
+            return result;
+        }
+
+        private string GetPath(Dictionary<string, OneDriveModel.ItemInfo> ids, OneDriveModel.OneDriveItem item)
+        {
+            var path = item.name;
+            var parentId = item.parentReference.id;
+            while (true)
+            {
+                OneDriveModel.ItemInfo info;
+                if (!ids.TryGetValue(parentId, out info))
+                {
+                    return path;
+                }
+
+                path = string.Format(CultureInfo.InvariantCulture, @"{0}\{1}", info.name, path);
+                parentId = info.parentId;
+            }
+        }
+
+        private async Task<ApplyResult> ApplyChangesParallel(OneDriveModel.OneDriveChangeCollection changes, string accessToken)
+        {
+            List<Task> tasks = new List<Task>();
+            string wwwroot = _environment.WebRootPath;
+            int failureCount = 0;
+            int successCount = 0;
+
+            using (var sem = new SemaphoreSlim(initialCount: System.Environment.ProcessorCount, maxCount: System.Environment.ProcessorCount))
+            using (var filesPerSecLimiter = new RateLimiter(MaxFilesPerSeconds, TimeSpan.FromSeconds(1)))
+            {
+                foreach (OneDriveModel.OneDriveChange change in changes)
+                {
+                    tasks.Add(Task.Run(async () =>
```

Comment:
Task.Run will not create new thread of each routine. it re-used thread from thread pool.
i think what we have should be fine.


````C#
            List<Task> tasks = new List<Task>();
            HashSet<int> ids = new HashSet<int>();
            int main = Thread.CurrentThread.ManagedThreadId;
            Console.WriteLine("Main Thread# {0}", main);
            for (int i = 0; i < 1500; i++)
            {
                tasks.Add(Task.Run(async () =>
                {
                    Random r = new Random();
                    await Task.Delay(TimeSpan.FromSeconds(r.Next(1, 5)));

                    int id = Thread.CurrentThread.ManagedThreadId;
                    lock (ids)
                    {
                        if (!ids.Contains(id))
                        {
                            ids.Add(id);
                        }
                    }
                }));
            }

            await Task.WhenAll(tasks);

            foreach (var item in ids)
            {
                Console.WriteLine(item);
            }

            int total = ids.Contains(main) ? ids.Count : ids.Count + 1;
            Console.WriteLine("Done! {0} threads used", total);
````

````
Main Thread# 3
11
7
3
13
6
9
5
8
12
4
16
10
14
Done! 13 threads used
Press any key to continue . . .
````